src/java/org/apache/cassandra/index/sai/disk/hnsw/CassandraOnDiskHnsw.java

```diff
@@ -99,33 +102,46 @@ public ReorderingPostingList search(float[] queryVector, int topK, Bits acceptBi
                                              view,
                                              ordinalsMap.ignoringDeleted(acceptBits),
                                              vistLimit);
-            return annRowIdsToPostings(queue);
+            return annRowIdsToPostings(queue, context, rowIdToPrimaryKeyMapper);
         }
         catch (IOException e)
         {
             throw new RuntimeException(e);
         }
     }
 
-    private class RowIdIterator implements PrimitiveIterator.OfInt, AutoCloseable
+    private class RowIdIteratorAndCacher implements PrimitiveIterator.OfInt, AutoCloseable
     {
         private final NeighborQueue queue;
+        private final QueryContext context;
+        private final RowIdToPrimaryKeyMapper mapper;
         private final OnDiskOrdinalsMap.RowIdsView rowIdsView = ordinalsMap.getRowIdsView();
 
         private PrimitiveIterator.OfInt segmentRowIdIterator = IntStream.empty().iterator();
 
-        public RowIdIterator(NeighborQueue queue)
+        public RowIdIteratorAndCacher(NeighborQueue queue, QueryContext context, RowIdToPrimaryKeyMapper mapper)
         {
+            assert mapper != null;
             this.queue = queue;
+            this.context = context;
+            this.mapper = mapper;
         }
 
         @Override
         public boolean hasNext() {
             while (!segmentRowIdIterator.hasNext() && queue.size() > 0) {
                 try
                 {
+                    var score = queue.topScore();
                     var ordinal = queue.pop();
-                    segmentRowIdIterator = Arrays.stream(rowIdsView.getSegmentRowIdsMatching(ordinal)).iterator();
+                    int[] rowIds = rowIdsView.getSegmentRowIdsMatching(ordinal);
+                    for (int rowId : rowIds)
+                    {
+                        PrimaryKey pk = mapper.getPrimaryKeyForRowId(rowId);
```

Comment:
this is probably a medium-sized refactor but could we push a PrimaryKeyMap instance into QueryContext?  that might  simplify a lot of passing PKM or PKMFactory around

Reply:
I will get vsearch-15 merged as a starting point

Reply:
I talked with @mike-tr-adamson about this briefly in slack. He agreed it would be more straight forward in some ways, but was concerned about always making sure it gets closed. From my perspective, I agree that it'd be a good place for it.

Do we want to make this contingent on that refactor?

Reply:
can you give it a try and see how it looks?  if it feels clunkier then we can skip it

Reply:
Sure

Reply:
Based on the slack discussion with @mike-tr-adamson, we will defer any refactoring. Re-open if I misunderstood.